k256/Cargo.toml

```diff
@@ -20,6 +20,7 @@ rust-version = "1.57"
 [dependencies]
 cfg-if = "1.0"
 elliptic-curve = { version = "0.12.3", default-features = false, features = ["hazmat", "sec1"] }
+lazy_static = { version = "1.4", features = ["spin_no_std"] }
```

Comment:
I generally prefer `once_cell` these days, especially as it's [slated for eventual inclusion in `core`/`std`](https://github.com/rust-lang/rust/issues/74465).

The `spin_no_std` "feature" of `lazy_static` was always weirdly non-additive too. Enabling it enables it for all crates that use `lazy_static`.

Reply:
Changed to `once_cell`. I am not 100% sure I'm using it correctly, so if you have experience with it, please check.

Reply:
It seems that to use `once_cell` in `no_std` environment we need the `critical_section` feature, but when such a library is used with `std` enabled, there is a linker error that can be prevented by specifying `critical-section = { version = "1.1", features = ["std"] }` explicitly. This is kind of a bummer. Is there a better way?

Reply:
Also `once_cell` has the MSRV 1.63, so this crate's MSRV will have to be updated.

Reply:
Huh? `once_cell` should be MSRV 1.56

> This is kind of a bummer. Is there a better way?

`no_std` users who want the basepoint tables can link `once_cell` and enable the `critical_section` feature themselves.

But I suspect many embedded will not basepoint tables at all due to limitations on the size of the loaded program. In that regard it would probably be good to feature gate them ala https://github.com/dalek-cryptography/curve25519-dalek/pull/489

Since you have a `mul_by_generator` function it can fallback to a slower implementation when the basepoint tables aren't available, but still provide a common API for either case.

Reply:
> Huh? once_cell should be MSRV 1.56

With no features, yes, but `critical-section` requires 1.63.

> Due to limitations on the size of the loaded program

It is a lazily evaluated value, so the limitation will be on the memory/cache usage. Although adding the evaluated lookup tables in the code itself (like `ed25519` does it) is another possibility, but it will still consume cache on usage.

> Since you have a mul_by_generator function it can fallback to a slower implementation when the basepoint tables aren't available, but still provide a common API for either case.

I thought the common API would be provided by the trait, which could use `mul_by_generator()` or `mul()` (in the default implementation)

Reply:
> no_std users who want the basepoint tables can link once_cell and enable the critical_section feature themselves.

Not sure I would like that as a user. Seems pretty brittle.

Reply:
> It is a lazily evaluated value, so the limitation will be on the memory/cache usage. 

Well yes, but that's what I mean by "size of the loaded program". It may need to fit in e.g. SRAM, on a device with no DRAM.

Lazy or not, at the end of the day it's a static table that needs to fit in whatever RAM the device has available, and many embedded devices are highly constrained.

As a data point, the `ed25519-dalek` basepoint tables were too big to fit into some ARM Cortex M0/M4 devices people were experimenting with at OxidizeConf when I attended it a few years ago.

> I thought the common API would be provided by the trait, which could use mul_by_generator() or mul() (in the default implementation)

We can use the trait when it's available, which provides a default `mul`-based implementation, and would let you conditionally override it in the impl block to use the basepoint tables.

But absent that upstream feature in `group`, you can do something like this with an inherent `mul_by_generator()` method:

https://github.com/dalek-cryptography/curve25519-dalek/pull/489/files#diff-0dbe5da3c59af6d6e20fbb1c92b248a7bcce0a5e15f0e5de6d16c9a60abe3ee3R705-R722

Reply:
> Not sure I would like that as a user. Seems pretty brittle.

Hopefully the ergonomic concerns here go away when it's merged upstream into `core`/`std`. Really that's the main reason to prefer it: it will soon hopefully be part of the standard library.

The `spin_no_std` feature of `lazy_static` is not a great solution here either. If one dependency in your `std`-dependent program activates it, suddenly every lazy static becomes spinlocked. And for that reason I think it's bad for dependencies to enable it rather than the toplevel `[bin]` crate.

Reply:
@fjarri another option is to transform the precomputed tables into Rust literal syntax, and use that to define a static. That's what the dalek crates do:

https://github.com/dalek-cryptography/curve25519-dalek/blob/3effd73307a606e44469a425974f0b7b0eb85899/src/backend/serial/u64/constants.rs#L328

You can test if it diverges by running the precomputation again and seeing if it equals the static value.

Reply:
That will lead to increased code size which may be less preferable than some start-up time, e.g. in WASM context.